pkg/resources/procedure.go

```diff
@@ -182,136 +185,380 @@ func Procedure() *schema.Resource {
+	runtimeVersion := d.Get("runtime_version").(string)
+
+	packages := []sdk.ProcedurePackageRequest{}
+	pkgs := d.Get("packages").([]interface{})
+	for _, pkg := range pkgs {
+		packages = append(packages, *sdk.NewProcedurePackageRequest(pkg.(string)))
+	}
+	handler := d.Get("handler").(string)
+	req := sdk.NewCreateForPythonProcedureRequest(schemaObjectId, *returns, runtimeVersion, packages, handler)
+	req.WithProcedureDefinition(sdk.String(procedureDefinition))
+	args, diags := getProcedureArguments(d)
+	if diags != nil {
+		return diags
+	}
+	if len(args) > 0 {
+		req.WithArguments(args)
+	}
+
+	// read optional params
+	if v, ok := d.GetOk("execute_as"); ok {
+		if strings.ToUpper(v.(string)) == "OWNER" {
+			req.WithExecuteAs(sdk.Pointer(sdk.ExecuteAsOwner))
+		} else if strings.ToUpper(v.(string)) == "CALLER" {
+			req.WithExecuteAs(sdk.Pointer(sdk.ExecuteAsCaller))
+		}
+	}
+	if v, ok := d.GetOk("null_input_behavior"); ok {
+		req.WithNullInputBehavior(sdk.Pointer(sdk.NullInputBehavior(v.(string))))
+	}
+	if v, ok := d.GetOk("comment"); ok {
+		req.WithComment(sdk.String(v.(string)))
+	}
+	if v, ok := d.GetOk("secure"); ok {
+		req.WithSecure(sdk.Bool(v.(bool)))
+	}
+	if _, ok := d.GetOk("imports"); ok {
+		imports := []sdk.ProcedureImportRequest{}
+		for _, imp := range d.Get("imports").([]interface{}) {
+			imports = append(imports, *sdk.NewProcedureImportRequest(imp.(string)))
+		}
+		req.WithImports(imports)
+	}
+
+	if err := client.Procedures.CreateForPython(ctx, req); err != nil {
+		return diag.FromErr(err)
 	}
-	rows, err := snowflake.Query(db, stmt)
+	argTypes := make([]string, len(args))
+	for i, arg := range args {
+		argTypes[i] = string(arg.ArgDataType)
+	}
+	id := NewProcedureID(database, schema, name, argTypes)
+	d.SetId(id.String())
+	return ReadContextProcedure(ctx, d, meta)
+}
+
+func ReadContextProcedure(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	diags := diag.Diagnostics{}
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	id, err := DecodeProcedureID(d.Id())
 	if err != nil {
-		return err
+		return diag.FromErr(err)
+	}
+	if err := d.Set("name", id.name.Name()); err != nil {
+		return diag.FromErr(err)
 	}
-	defer rows.Close()
-	descPropValues, err := snowflake.ScanProcedureDescription(rows)
+	if err := d.Set("database", id.DatabaseName()); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("schema", id.SchemaName()); err != nil {
+		return diag.FromErr(err)
+	}
+	args := d.Get("arguments").([]interface{})
+	argTypes := make([]string, len(args))
+	for i, arg := range args {
+		argTypes[i] = arg.(map[string]interface{})["type"].(string)
+	}
+	procedureDetails, err := client.Procedures.Describe(ctx, sdk.NewDescribeProcedureRequest(id.name, id.argTypes))
 	if err != nil {
-		return err
+		// if procedure is not found then mark resource to be removed from state file during apply or refresh
+		d.SetId("")
+		return nil
```

Comment:
maybe at least we should return a warning that Describe failed?

Reply:
added warning